packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -1108,43 +1137,58 @@ export function simplify<T>(
 					const key = typeof field.key === 'string'
 						? field.key
 						: simplify(field.key, options, parents)
-					if (typeof key !== 'string' && key.kind === 'union') {
-						for (const subKey of key.members) {
-							fields.push({
-								...field,
-								key: simplifyKey(subKey),
-							})
-						}
-					} else {
-						fields.push({
-							...field,
-							key: simplifyKey(key),
-						})
-					}
+					addField(key, field)
 				} else {
 					const simplifiedSpreadType = simplify(
 						field.type,
 						options,
 						parents,
 					)
-
 					if (simplifiedSpreadType.kind === 'struct') {
-						fields.push(...simplifiedSpreadType.fields)
+						for (const field of simplifiedSpreadType.fields) {
+							addField(field.key, field)
+						}
+					} else {
+						const type = McdocType.toString(simplifiedSpreadType)
+						options.context.logger.warn(
+							`Tried to spread non-struct type ${type}`,
+						)
 					}
 				}
 			}
+			complexFields = complexFields.filter(other =>
+				!isAssignable(
+					typeof other.key === 'string'
+						? {
+							kind: 'literal',
+							value: { kind: 'string', value: other.key },
+						}
+						: other.key,
+					{
+						kind: 'union',
+						members: [...literalFields.keys()].map(k => ({
+							kind: 'literal',
+							value: { kind: 'string', value: k },
+						})),
+					},
+					options.context,
+					options.isEquivalent,
+				)
+			)
 			return {
 				kind: 'struct',
-				fields: fields.filter((f, i) =>
-					fields.findLastIndex(of =>
-						isAssignable(
-							of.key,
-							f.key,
-							options.context,
-							options.isEquivalent,
-						)
-					) === i
-				),
+				fields: [
+					...complexFields,
+					...[...literalFields.entries()].map(([key, field]) =>
```

Comment:
why do we flatten the `Map` we just made instead of returning it as-is? like

```suggestion
				fields: {
					complexFields,
					literalFields,
				}
```

and re-using it later...?

by flattening it, it looks like we're just gonna have to redo the `addField` stuff again later

Reply:
Because some of the complex fields are not represented in a map. It is true that nested struct spreading will cause the literal fields to be converted between a map and a list a few times. Maybe this should be improved. But simply returning a map here is not possible.

Reply:
Yeah, this would need some more rework to facilitate this, and the performance issue we ran into is not really related to this.